**Capstone Project Module #2 - Sendhi Anshari Rasyid**

# **Latar Belakang**

# **Pernyataan Masalah**

1. Faktor apa saja yang mempengaruhi kunjungan dari sebuah penginapan? | *harga, lokasi, jumlah minimal hari menginap, tipe penginapan*
1. Kapankah suatu penginapan ramai dikunjungi oleh para penyewa?
1. Apakah persebaran penginapan sudah cukup merata di Bangkok?

# **Tujuan**

1. Mengetahui faktor yang paling berpengaruh terhadap kunjungan sebuah penginapan di Bangkok
1. Mengetahui bulan apa yang menjadi puncak dari kunjungan pada penginapan di Bangkok
1. Mengetahui daerah mana yang menjadi pusat kunjungan dari para pendatang yang singgah di Bangkok

## **Data**
Dataset yang digunakan pada analisis ini merupakan kumpulan dari *host* penginapan dalam aplikasi Airbnb yang berlokasi di sekitar Kota Bangkok, Thailand. Lengkap dengan detail lokasi, ketersediaan kamar, tipe kamar, dan lainnya. Dataset tersebut bisa diakses melalui [tautan ini.](https://drive.google.com/file/d/1Kagt-IMGruvyBV3tH6HYa721JK-TN-56/view?usp=drive_link)

### Penjelasan Dataset

Berikut adalah penjelasan detail dari setiap data yang termasuk dalam dataset tersebut:

-  `Unnamed: 0`: Index
-  `id`: ID Airbnb unik yang dimiliki setiap listing penginapan
-  `name`: Nama listing penginapan
-  `host_id`: ID Airbnb unik yang dimiliki oleh setiap host
-  `host_name`: Nama dari host pemilik penginapan
-  `neighborhood`: Lokasi/daerah di mana penginapan berada
-  `latitude`: Posisi koordinat garis lintang dari penginapan terkait
-  `longitude`: Posisi koordinat garis bujur dari penginapan terkait
-  `room_type`: Tipe dari kamar yang disewakan, terbagi menjadi 3; *entire place, private rooms, shared rooms*
-  `price`: Harga sewa per hari
-  `minimum_nights`: Jumlah minimal untuk menyewa kamar
-  `number_of_reviews`: Jumlah review keseluruhan yang dimiliki oleh penginapan
-  `last_review`: Tanggal review terakhir oleh konsumen
-  `reviews_per_month`: Jumlah review yang diterima oleh penginapan pada satu bulan
-  `calculated_host_listings_count`: Jumlah listing yang dimiliki oleh setiap host
-  `availability_365`: Ketersediaan kamar yang dapat disewa dalam kurun waktu 365 hari
-  `number_of_reviews_ltm`: Jumlah review yang dimiliki oleh penginapan (dalam 12 bulan terakhir)

### Data Preparation
Tahap persiapan dengan melakukan *import* beberapa *library* yang akan digunakan dalam mengelola dan menganalisis data lebih lanjut.

In [124]:
#Import Library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from scipy.stats import normaltest, chi2_contingency
import warnings
warnings.filterwarnings('ignore')

In [125]:
#Import Dataset
df=pd.read_csv('Airbnb Listings Bangkok.csv')

#Menampilkan 5 data sample acak yang terdapat di dalam dataset
display(df.sample(5))

Unnamed: 0                  id  \
14060  14060       715565949199588020   
13793  13793       689719738192889434   
14624  14624       743694980810006625   
7779   7779        34289227             
4470   4470        22720798             

                                                    name    host_id host_name  \
14060  Lovely Studio room with pool - great location      434168399  Krittika   
13793  Modern Cozy fully furnished 1 BR | Center of Bkk   474141399  Beverly    
14624  素坤逸高级公寓近BTS on nut，无边天际泳池赏曼谷夜景                     484194134  行走在曼谷      
7779   Luxury condo@Superb skyline view                   199624792  Ho         
4470   1Br Silom-Sathorn Bkk@BTS Chong Nonsi by Triple B  164560603  Triple B   

      neighbourhood  latitude  longitude        room_type  price  \
14060  Khlong Toei   13.72693  100.56259  Entire home/apt  1950    
13793  Khlong Toei   13.70901  100.58296  Entire home/apt  797     
14624  Khlong Toei   13.72900  100.56854  Private room     2500    
7779   Vadhana       13.74019  100.58869  Entire home/apt  1450    
4470   Bang Rak      13.72360  100.52676  Entire home/apt  2429    

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
14060  1               0                  NaN        NaN                  
13793  14              5                  2022-12-13  1.14                
14624  28              0                  NaN        NaN                  
7779   21              2                  2022-04-01  0.21                
4470   14              36                 2022-12-22  0.61                

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
14060  44                              359               0                      
13793  3                               331               5                      
14624  1                               362               0                      
7779   23                              95                2                      
4470   3                               320               5

> Terlihat bahwa kolom `Unnamed: 0` bertindak sebagai index dari setiap datanya. Untuk menghindari kebingungan, maka kolom tersebut akan diubah namanya menjadi kolom `Index`.

In [126]:
#Mengganti nama kolom
df=df.rename(columns={'Unnamed: 0':'index'})
df.head(5)

index      id                                            name  host_id  \
0  0      27934   Nice room with superb city view                 120437    
1  1      27979   Easy going landlord,easy place                  120541    
2  2      28745   modern-style apartment in Bangkok               123784    
3  3      35780   Spacious one bedroom at The Kris Condo Bldg. 3  153730    
4  4      941865  Suite Room 3 at MetroPoint                      610315    

    host_name neighbourhood  latitude  longitude        room_type  price  \
0  Nuttee      Ratchathewi   13.75983  100.54134  Entire home/apt  1905    
1  Emy         Bang Na       13.66818  100.61674  Private room     1316    
2  Familyroom  Bang Kapi     13.75232  100.62402  Private room     800     
3  Sirilak     Din Daeng     13.78823  100.57256  Private room     1286    
4  Kasem       Bang Kapi     13.76872  100.63338  Private room     1905    

   minimum_nights  number_of_reviews last_review  reviews_per_month  \
0  3               65                 2020-01-06  0.50                
1  1               0                  NaN        NaN                  
2  60              0                  NaN        NaN                  
3  7               2                  2022-04-01  0.03                
4  1               0                  NaN        NaN                  

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  
0  2                               353               0                      
1  2                               358               0                      
2  1                               365               0                      
3  1                               323               1                      
4  3                               365               0

### Data Understanding dan Data Cleaning
Sebelum melakukan analisis, terlebih dahulu dilakukan *data understanding* untuk lebih memahami isi dataset. Serta *data cleaning* untuk mengantisipasi kehadiran *error* ataupun *missing value* pada sebuah dataset.

Kedua hal tersebut dilakukan agar *data analysis* yang dilakukan nantinya bisa lebih efisien dan memiliki hasil yang maksimal.

Adapun beberapa tahapan tahapan yang akan saya lakukan, terdiri dari:

1. Identifikasi Data
1. Identifikasi Kolom yang Akan Dianalisis
1. Identifikasi Duplikasi Data pada Primary Key
1. Identifikasi dan Penanganan *Missing Value*
1. Identifikasi *Unique Data*

##### 1. Identifikasi Data

In [127]:
#Mengetahui jumlah data yang terdapat pada dataset
a,b = df.shape
print(f'Jumlah data yang terdapat pada adalah sebanyak {a} baris dan {b} kolom.\n')

#Melihat detail dataset
df.info()

Jumlah data yang terdapat pada adalah sebanyak 15854 baris dan 17 kolom.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15854 entries, 0 to 15853
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   index                           15854 non-null  int64  
 1   id                              15854 non-null  int64  
 2   name                            15846 non-null  object 
 3   host_id                         15854 non-null  int64  
 4   host_name                       15853 non-null  object 
 5   neighbourhood                   15854 non-null  object 
 6   latitude                        15854 non-null  float64
 7   longitude                       15854 non-null  float64
 8   room_type                       15854 non-null  object 
 9   price                           15854 non-null  int64  
 10  minimum_nights                  15854 non-null  int64  
 11  number_of_reviews  

> Beberapa informasi yang bisa didapatkan berdasarkan data di atas adalah sebagai berikut:
> - Jumlah data yang terdapat pada dataset ini adalah sebanyak 15854 baris yang terbagi pada 17 kolom.
>- Jumlah data pada kolom `last_review` dan `reviews_per_month` hanya sebanyak 10064 data saja. Jauh berbeda dengan total jumlah data keseluruhan, yaitu sebanyak 15854 data. Dapat disimpulkan (sementara) bahwa pada kedua kolom tersebut terdapat banyak data *Null*.
>- Tipe data yang terdapat di dalam dataset terbagi menjadi 12 kolom numerik dan 5 kolom non-numerik/object.
>- Format pada kolom `last_review` masih keliru, seharusnya berformat *date.*

In [128]:
#Mengubah format 'last_review' menjadi datetime
df['last_review'] = pd.to_datetime(df['last_review'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15854 entries, 0 to 15853
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   index                           15854 non-null  int64         
 1   id                              15854 non-null  int64         
 2   name                            15846 non-null  object        
 3   host_id                         15854 non-null  int64         
 4   host_name                       15853 non-null  object        
 5   neighbourhood                   15854 non-null  object        
 6   latitude                        15854 non-null  float64       
 7   longitude                       15854 non-null  float64       
 8   room_type                       15854 non-null  object        
 9   price                           15854 non-null  int64         
 10  minimum_nights                  15854 non-null  int64         
 11  nu

> Kolom `reviews_per_month` sudah berhasil diganti tipe datanya menjadi datetime.

In [129]:
#Melihat statistik deskriptif dari masing-masing kolom data (numerik dan non-numerik)
display(df.describe(), df.describe(include='object'))

index            id       host_id      latitude     longitude  \
count  15854.000000  1.585400e+04  1.585400e+04  15854.000000  15854.000000   
mean   7926.500000   1.579397e+17  1.541058e+08  13.745144     100.559903     
min    0.000000      2.793400e+04  5.892000e+04  13.527300     100.329550     
25%    3963.250000   2.104509e+07  3.974431e+07  13.720090     100.529690     
50%    7926.500000   3.503734e+07  1.224556e+08  13.738490     100.561415     
75%    11889.750000  5.256154e+07  2.390547e+08  13.759497     100.585150     
max    15853.000000  7.908162e+17  4.926659e+08  13.953540     100.923440     
std    4576.799919   2.946015e+17  1.318726e+08  0.043040      0.050911       

              price  minimum_nights  number_of_reviews  \
count  1.585400e+04  15854.000000    15854.000000        
mean   3.217704e+03  15.292355       16.654157           
min    0.000000e+00  1.000000        0.000000            
25%    9.000000e+02  1.000000        0.000000            
50%    1.429000e+03  1.000000        2.000000            
75%    2.429000e+03  7.000000        13.000000           
max    1.100000e+06  1125.000000     1224.000000         
std    2.497212e+04  50.815020       40.613331           

                         last_review  reviews_per_month  \
count  10064                          10064.000000        
mean   2021-08-30 08:37:49.316375296  0.813145            
min    2012-12-15 00:00:00            0.010000            
25%    2020-02-20 00:00:00            0.120000            
50%    2022-10-24 00:00:00            0.435000            
75%    2022-12-08 00:00:00            1.060000            
max    2022-12-28 00:00:00            19.130000           
std    NaN                            1.090196            

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
count  15854.000000                    15854.000000      15854.000000           
mean   13.889618                       244.378643        3.481519               
min    1.000000                        0.000000          0.000000               
25%    1.000000                        138.000000        0.000000               
50%    4.000000                        309.000000        0.000000               
75%    13.000000                       360.000000        3.000000               
max    228.000000                      365.000000        325.000000             
std    30.269848                       125.843224        8.916937

name host_name  \
count   15846                                             15853      
unique  14794                                             5312       
top     New!  La Chada Night Market studio 2PPL near MRT  Curry      
freq    45                                                228        

       neighbourhood        room_type  
count   15854         15854            
unique  50            4                
top     Vadhana       Entire home/apt  
freq    2153          8912

> Berdasarkan informasi di atas, kita bisa mengetahui beberapa hal:
> - Pada kolom `Price` diketahui memiliki nilai minimal sebesar 0, di mana merupakan sebuah anomali. Karena *value* tersebut mengindikasikan bahwa ada sebuah *listing* penginapan yang bisa disewa dengan gratis. Sedangkan sebuah *listing* seharusnya memiliki *value* > 0 agar bisa disewakan.
> - Jumlah data *unique* pada kolom `name`, `host_name`, `neighbourhood`, `room_type`, dan `last_review` memiliki selisih yang signifikan dengan total data *(count)* keseluruhan, sehingga dapat disimpulkan terdapat banyak duplikasi data. Hal ini bukanlah suatu anomali, mengingat kolom-kolom tersebut memang memiliki *value* yang berulang.

In [130]:
#Melihat data yang memiliki 'price' = 0
df.loc[df['price'] == 0]

index        id                           name    host_id  \
11103  11103  44563108  Somerset Maison Asoke Bangkok  360620448   

                   host_name neighbourhood  latitude  longitude   room_type  \
11103  Somerset Maison Asoke  Vadhana       13.73815  100.5642   Hotel room   

       price  minimum_nights  number_of_reviews last_review  \
11103  0      1               0                 NaT           

       reviews_per_month  calculated_host_listings_count  availability_365  \
11103 NaN                 1                               0                  

       number_of_reviews_ltm  
11103  0

> Diketahui terdapat 1 data yang memiliki `price` = 0. Karena merupakan sebuah anomali data, maka diputuskan untuk menghapus row data tersebut.

In [131]:
#Menghapus row data yang memiliki 'price' = 0
df.drop(11103, inplace=True)

In [132]:
#Reset index setelah melakukan drop row
df.set_index('name', inplace=True)
df.reset_index(inplace=True)

In [133]:
#Pengecekan ulang statistik deskriptif dari masing-masing kolom data
display(df.describe(), df.describe(include='object'))

index            id       host_id      latitude     longitude  \
count  15853.000000  1.585300e+04  1.585300e+04  15853.000000  15853.000000   
mean   7926.299628   1.579496e+17  1.540928e+08  13.745144     100.559903     
min    0.000000      2.793400e+04  5.892000e+04  13.527300     100.329550     
25%    3963.000000   2.104509e+07  3.974431e+07  13.720090     100.529690     
50%    7926.000000   3.503224e+07  1.224556e+08  13.738490     100.561410     
75%    11890.000000  5.256284e+07  2.390274e+08  13.759500     100.585150     
max    15853.000000  7.908162e+17  4.926659e+08  13.953540     100.923440     
std    4576.874737   2.946082e+17  1.318666e+08  0.043041      0.050912       

              price  minimum_nights  number_of_reviews  \
count  1.585300e+04  15853.000000    15853.000000        
mean   3.217907e+03  15.293257       16.655207           
min    2.780000e+02  1.000000        0.000000            
25%    9.000000e+02  1.000000        0.000000            
50%    1.429000e+03  1.000000        2.000000            
75%    2.429000e+03  7.000000        13.000000           
max    1.100000e+06  1125.000000     1224.000000         
std    2.497290e+04  50.816496       40.614397           

                         last_review  reviews_per_month  \
count  10064                          10064.000000        
mean   2021-08-30 08:37:49.316375296  0.813145            
min    2012-12-15 00:00:00            0.010000            
25%    2020-02-20 00:00:00            0.120000            
50%    2022-10-24 00:00:00            0.435000            
75%    2022-12-08 00:00:00            1.060000            
max    2022-12-28 00:00:00            19.130000           
std    NaN                            1.090196            

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
count  15853.000000                    15853.000000      15853.000000           
mean   13.890431                       244.394058        3.481738               
min    1.000000                        0.000000          0.000000               
25%    1.000000                        138.000000        0.000000               
50%    4.000000                        309.000000        0.000000               
75%    13.000000                       360.000000        3.000000               
max    228.000000                      365.000000        325.000000             
std    30.270630                       125.832224        8.917176

name host_name  \
count   15845                                             15852      
unique  14793                                             5311       
top     New!  La Chada Night Market studio 2PPL near MRT  Curry      
freq    45                                                228        

       neighbourhood        room_type  
count   15853         15853            
unique  50            4                
top     Vadhana       Entire home/apt  
freq    2152          8912

> Dapat terlihat bahwa nilai min dari `price` sudah bukan 0 lagi, yang berarti data '11103' yang memiliki value=0 sudah berhasil terhapus.

##### 2. Identifikasi Kolom yang Akan Dianalisis

Sebelum memasuki tahapan analisis data, terlebih dahulu mengidentifikasi kolom mana yang tidak akan digunakan dalam tahap tersebut karena tidak relevan dengan rumusan masalah. Beberapa yang saya putuskan untuk dihapus secara permanen adalah sebagai berikut:
- kolom `Index`, karena informasinya hanya merupakan duplikasi dari index setiap data
- kolom `calculated_host_listings_count`, karena hanya berisikan informasi terkait jumlah penginapan yang dimiliki oleh seorang host
- kolom `number_of_reviews_ltm`, karena hanya merupakan rangkuman dari jumlah review yang diterima dalam 12 bulan terakhir
- kolom `longitude` dan `latitude`, karena hanya berisikan koordinat detail dari setiap penginapannya
- kolom `reviews_per_month`, karena hanya berisi jumlah review yang diterima oleh sebuah penginapan dalam setiap bulannya

In [134]:
# df.drop(['index', 'longitude', 'latitude', 'reviews_per_month', 'calculated_host_listings_count', 'number_of_reviews_ltm'], axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15853 entries, 0 to 15852
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   name                            15845 non-null  object        
 1   index                           15853 non-null  int64         
 2   id                              15853 non-null  int64         
 3   host_id                         15853 non-null  int64         
 4   host_name                       15852 non-null  object        
 5   neighbourhood                   15853 non-null  object        
 6   latitude                        15853 non-null  float64       
 7   longitude                       15853 non-null  float64       
 8   room_type                       15853 non-null  object        
 9   price                           15853 non-null  int64         
 10  minimum_nights                  15853 non-null  int64         
 11  nu

##### 3. Identifikasi Duplikasi Data pada Primary Key

Karena kolom `ID` bertindak sebagai *primary key* pada dataset ini, maka prioritas pertama adalah mengecek apakah kolom tersebut memiliki duplikasi atau tidak agar akurasi dari analisis yang dilakukan bisa lebih akurat.

In [135]:
#Pengecekan duplikasi data pada kolom ID
print("Jumlah duplikasi data pada kolom 'ID' adalah sebanyak",
      df['id'].duplicated().sum(), "data.")

Jumlah duplikasi data pada kolom 'ID' adalah sebanyak 0 data.


> Setelah dilakukan pengecekan, ternyata kolom `ID` tidak memiliki duplikasi. 

Pada kolom lainnya tidak dicek apakah memiliki duplikasi atau tidak, karena sudah terwakili oleh pengecekan kolom `ID` ini. Kemunculan data berulang yang terjadi pada kolom `host_name`, `host_id`, `neighbourhood`, ataupun lainnya tidak akan berpengaruh pada proses analisis berikutnya.

##### 4. Identifikasi dan Penanganan Missing Value

In [136]:
#Fungsi tampil missing value
def persentase_null():
    return round(df.isnull().sum() * 100 / len(df), 2)

def jumlah_null():
    return df.isna().sum()

#Dataframe missing value
df_null = pd.DataFrame({
    'Jumlah Data': jumlah_null(),
    'Persentase': persentase_null()
})

#Pengecekan keberadaan missing value dari setiap kolom
print('Jumlah missing value di setiap kolom adalah sebagai berikut:\n')
print(df_null)
print('\nJumlah missing value dari semua kolom sebanyak', jumlah_null().sum(), 'data.')

Jumlah missing value di setiap kolom adalah sebagai berikut:

                                Jumlah Data  Persentase
name                            8            0.05      
index                           0            0.00      
id                              0            0.00      
host_id                         0            0.00      
host_name                       1            0.01      
neighbourhood                   0            0.00      
latitude                        0            0.00      
longitude                       0            0.00      
room_type                       0            0.00      
price                           0            0.00      
minimum_nights                  0            0.00      
number_of_reviews               0            0.00      
last_review                     5789         36.52     
reviews_per_month               5789         36.52     
calculated_host_listings_count  0            0.00      
availability_365                0         

>1. Beberapa kolom masih memiliki *missing value*, dengan detail sebagai berikut:
>- `name` sebanyak 8 data atau 0.05%
>- `host_name` sebanyak 1 data atau 0.01%
>- `last_review` sebanyak 5790 data atau 36.52%
>- `reviews_per_month` sebanyak 5790 data atau 36.52%
>2. Total dari keseluruhan *missing value* yang terdapat pada dataset adalah sejumlah 11589 data.
>3. Walaupun memiliki *missing value* yang cukup besar presentasinya, karena kolom `last_review` akan dijadikan sebagai salah satu acuan dalam melakukan analisis, maka kolom tersebut tetap dipertahankan.

Kolom-kolom yang masih memiliki *missing value* tersebut akan dicek satu per satu, sebelum diputuskan tindak lanjut seperti apa yang harus dilakukan pada data-data tersebut.

In [137]:
#Pengecekan missing value pada kolom name
df[df['name'].isna()]

name  index        id   host_id        host_name neighbourhood  latitude  \
439   NaN  439    4549768   18852579  Titawan          Phra Khanong  13.69406   
544   NaN  544    4720818   24386225  Cherry           Din Daeng     13.77562   
572   NaN  572    4245018   22030043  Parichart        Bang Phlat    13.78376   
669   NaN  669    6148415   31895202  Chira            Bang Na       13.68276   
1030  NaN  1030   8055144   42521288  Nantida          Vadhana       13.74126   
1282  NaN  1282   10000742  51374914  Diamond Bangkok  Ratchathewi   13.75328   
1594  NaN  1594   10710165  55347997  Khaneungnit      Vadhana       13.71757   
2075  NaN  2075   13142743  73275200  Pakaphol         Khlong Toei   13.72566   

      longitude        room_type  price  minimum_nights  number_of_reviews  \
439   100.59619  Private room     1080   5               0                   
544   100.57346  Private room     1200   1               0                   
572   100.49821  Private room     1200   1               0                   
669   100.60894  Entire home/apt  2424   2               0                   
1030  100.55761  Private room     5000   3               0                   
1282  100.52928  Private room     930    1               6                   
1594  100.60464  Private room     1000   1               0                   
2075  100.56416  Private room     850    1               2                   

     last_review  reviews_per_month  calculated_host_listings_count  \
439  NaT         NaN                 1                                
544  NaT         NaN                 1                                
572  NaT         NaN                 1                                
669  NaT         NaN                 1                                
1030 NaT         NaN                 1                                
1282 2017-05-13   0.07               1                                
1594 NaT         NaN                 1                                
2075 2017-12-11   0.03               3                                

      availability_365  number_of_reviews_ltm  
439   365               0                      
544   365               0                      
572   365               0                      
669   365               0                      
1030  365               0                      
1282  365               0                      
1594  365               0                      
2075  220               0

In [138]:
#Proses mengubah missing value pada kolom `name` menjadi "No Name"
df['name'].fillna('No Name', inplace=True)
df[df['name'] == 'No Name']

name  index        id   host_id        host_name neighbourhood  \
439   No Name  439    4549768   18852579  Titawan          Phra Khanong   
544   No Name  544    4720818   24386225  Cherry           Din Daeng      
572   No Name  572    4245018   22030043  Parichart        Bang Phlat     
669   No Name  669    6148415   31895202  Chira            Bang Na        
1030  No Name  1030   8055144   42521288  Nantida          Vadhana        
1282  No Name  1282   10000742  51374914  Diamond Bangkok  Ratchathewi    
1594  No Name  1594   10710165  55347997  Khaneungnit      Vadhana        
2075  No Name  2075   13142743  73275200  Pakaphol         Khlong Toei    

      latitude  longitude        room_type  price  minimum_nights  \
439   13.69406  100.59619  Private room     1080   5                
544   13.77562  100.57346  Private room     1200   1                
572   13.78376  100.49821  Private room     1200   1                
669   13.68276  100.60894  Entire home/apt  2424   2                
1030  13.74126  100.55761  Private room     5000   3                
1282  13.75328  100.52928  Private room     930    1                
1594  13.71757  100.60464  Private room     1000   1                
2075  13.72566  100.56416  Private room     850    1                

      number_of_reviews last_review  reviews_per_month  \
439   0                 NaT         NaN                  
544   0                 NaT         NaN                  
572   0                 NaT         NaN                  
669   0                 NaT         NaN                  
1030  0                 NaT         NaN                  
1282  6                 2017-05-13   0.07                
1594  0                 NaT         NaN                  
2075  2                 2017-12-11   0.03                

      calculated_host_listings_count  availability_365  number_of_reviews_ltm  
439   1                               365               0                      
544   1                               365               0                      
572   1                               365               0                      
669   1                               365               0                      
1030  1                               365               0                      
1282  1                               365               0                      
1594  1                               365               0                      
2075  3                               220               0

> Seluruh *missing value* pada kolom `name` kini telah diisi oleh *value* "No Name."

In [139]:
#Pengecekan missing value pada kolom host name
df[df['host_name'].isna()]

name  index        id    host_id host_name neighbourhood  \
3571  Cozy Hideaway  3571   19682464  137488762  NaN       Bang Kapi      

      latitude  longitude     room_type  price  minimum_nights  \
3571  13.76999  100.63769  Private room  1399   3                

      number_of_reviews last_review  reviews_per_month  \
3571  1                 2017-07-29   0.02                

      calculated_host_listings_count  availability_365  number_of_reviews_ltm  
3571  1                               365               0

In [140]:
#Proses mengubah missing value pada kolom `name` menjadi "No Name"
df['host_name'].fillna('No Name', inplace=True)
df[df['host_name'] == 'No Name']

name  index        id    host_id host_name neighbourhood  \
3571  Cozy Hideaway  3571   19682464  137488762  No Name   Bang Kapi      

      latitude  longitude     room_type  price  minimum_nights  \
3571  13.76999  100.63769  Private room  1399   3                

      number_of_reviews last_review  reviews_per_month  \
3571  1                 2017-07-29   0.02                

      calculated_host_listings_count  availability_365  number_of_reviews_ltm  
3571  1                               365               0

> *Missing value* pada kolom `host_name` kini telah diisi oleh *value* "No Name."

In [141]:
#Pengecekan missing value pada kolom last_review
df[df['last_review'].isna()]

name  index  \
1      Easy going landlord,easy place                      1       
2      modern-style apartment in Bangkok                   2       
4      Suite Room 3 at MetroPoint                          4       
7      1 chic bedroom apartment in BKK                     7       
14     Deluxe Condo, Nana, Pool/GYM/Sauna                  14      
...                                   ...                  ..      
15848  素坤逸核心两房公寓42楼，靠近BTSon nut/无边天际泳池观赏曼谷夜景/出门当地美食街       15849   
15849  Euro LuxuryHotel PratunamMKt TripleBdNrShopingArea  15850   
15850  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15851   
15851  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15852   
15852  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15853   

                       id    host_id   host_name neighbourhood   latitude  \
1      27979               120541     Emy         Bang Na       13.668180   
2      28745               123784     Familyroom  Bang Kapi     13.752320   
4      941865              610315     Kasem       Bang Kapi     13.768720   
7      1738669             7045870    Jiraporn    Chatu Chak    13.829250   
14     959254              5153476    Natcha      Khlong Toei   13.715160   
...       ...                  ...       ...              ...         ...   
15848  790465040741092826  94899359   Renee       Pra Wet       13.715132   
15849  790474503157243541  491526222  Phakhamon   Ratchathewi   13.753052   
15850  790475335086864240  491526222  Phakhamon   Ratchathewi   13.753169   
15851  790475546213717328  491526222  Phakhamon   Ratchathewi   13.754789   
15852  790476492384199044  491526222  Phakhamon   Ratchathewi   13.752960   

        longitude        room_type  price  minimum_nights  number_of_reviews  \
1      100.616740  Private room     1316   1               0                   
2      100.624020  Private room     800    60              0                   
4      100.633380  Private room     1905   1               0                   
7      100.567370  Entire home/apt  1461   1               0                   
14     100.568060  Entire home/apt  1400   30              0                   
...           ...              ...   ...   ..             ..                   
15848  100.653458  Private room     2298   28              0                   
15849  100.538738  Private room     1429   1               0                   
15850  100.538700  Private room     1214   1               0                   
15851  100.538757  Private room     1214   1               0                   
15852  100.540820  Private room     1214   1               0                   

      last_review  reviews_per_month  calculated_host_listings_count  \
1     NaT         NaN                 2                                
2     NaT         NaN                 1                                
4     NaT         NaN                 3                                
7     NaT         NaN                 1                                
14    NaT         NaN                 1                                
...    ..          ..                ..                                
15848 NaT         NaN                 1                                
15849 NaT         NaN                 14                               
15850 NaT         NaN                 14                               
15851 NaT         NaN                 14                               
15852 NaT         NaN                 14                               

       availability_365  number_of_reviews_ltm  
1      358               0                      
2      365               0                      
4      365               0                      
7      365               0                      
14     365               0                      
...    ...              ..                      
15848  362               0                      
15849  365               0                      
15850  365               0                  

In [142]:
#Proses mengubah missing value pada kolom `last_review` menjadi "-"
df['last_review'].fillna('-', inplace=True)
df[df['last_review'] == '-']

name  index  \
1      Easy going landlord,easy place                      1       
2      modern-style apartment in Bangkok                   2       
4      Suite Room 3 at MetroPoint                          4       
7      1 chic bedroom apartment in BKK                     7       
14     Deluxe Condo, Nana, Pool/GYM/Sauna                  14      
...                                   ...                  ..      
15848  素坤逸核心两房公寓42楼，靠近BTSon nut/无边天际泳池观赏曼谷夜景/出门当地美食街       15849   
15849  Euro LuxuryHotel PratunamMKt TripleBdNrShopingArea  15850   
15850  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15851   
15851  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15852   
15852  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15853   

                       id    host_id   host_name neighbourhood   latitude  \
1      27979               120541     Emy         Bang Na       13.668180   
2      28745               123784     Familyroom  Bang Kapi     13.752320   
4      941865              610315     Kasem       Bang Kapi     13.768720   
7      1738669             7045870    Jiraporn    Chatu Chak    13.829250   
14     959254              5153476    Natcha      Khlong Toei   13.715160   
...       ...                  ...       ...              ...         ...   
15848  790465040741092826  94899359   Renee       Pra Wet       13.715132   
15849  790474503157243541  491526222  Phakhamon   Ratchathewi   13.753052   
15850  790475335086864240  491526222  Phakhamon   Ratchathewi   13.753169   
15851  790475546213717328  491526222  Phakhamon   Ratchathewi   13.754789   
15852  790476492384199044  491526222  Phakhamon   Ratchathewi   13.752960   

        longitude        room_type  price  minimum_nights  number_of_reviews  \
1      100.616740  Private room     1316   1               0                   
2      100.624020  Private room     800    60              0                   
4      100.633380  Private room     1905   1               0                   
7      100.567370  Entire home/apt  1461   1               0                   
14     100.568060  Entire home/apt  1400   30              0                   
...           ...              ...   ...   ..             ..                   
15848  100.653458  Private room     2298   28              0                   
15849  100.538738  Private room     1429   1               0                   
15850  100.538700  Private room     1214   1               0                   
15851  100.538757  Private room     1214   1               0                   
15852  100.540820  Private room     1214   1               0                   

      last_review  reviews_per_month  calculated_host_listings_count  \
1      -          NaN                 2                                
2      -          NaN                 1                                
4      -          NaN                 3                                
7      -          NaN                 1                                
14     -          NaN                 1                                
...   ..           ..                ..                                
15848  -          NaN                 1                                
15849  -          NaN                 14                               
15850  -          NaN                 14                               
15851  -          NaN                 14                               
15852  -          NaN                 14                               

       availability_365  number_of_reviews_ltm  
1      358               0                      
2      365               0                      
4      365               0                      
7      365               0                      
14     365               0                      
...    ...              ..                      
15848  362               0                      
15849  365               0                      
15850  365               0                  

> *Missing value* pada kolom `last_review` kini telah diisi oleh *value* "-"

In [143]:
#Pengecekan missing value pada kolom reviews_per_month
df[df['reviews_per_month'].isna()]

name  index  \
1      Easy going landlord,easy place                      1       
2      modern-style apartment in Bangkok                   2       
4      Suite Room 3 at MetroPoint                          4       
7      1 chic bedroom apartment in BKK                     7       
14     Deluxe Condo, Nana, Pool/GYM/Sauna                  14      
...                                   ...                  ..      
15848  素坤逸核心两房公寓42楼，靠近BTSon nut/无边天际泳池观赏曼谷夜景/出门当地美食街       15849   
15849  Euro LuxuryHotel PratunamMKt TripleBdNrShopingArea  15850   
15850  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15851   
15851  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15852   
15852  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15853   

                       id    host_id   host_name neighbourhood   latitude  \
1      27979               120541     Emy         Bang Na       13.668180   
2      28745               123784     Familyroom  Bang Kapi     13.752320   
4      941865              610315     Kasem       Bang Kapi     13.768720   
7      1738669             7045870    Jiraporn    Chatu Chak    13.829250   
14     959254              5153476    Natcha      Khlong Toei   13.715160   
...       ...                  ...       ...              ...         ...   
15848  790465040741092826  94899359   Renee       Pra Wet       13.715132   
15849  790474503157243541  491526222  Phakhamon   Ratchathewi   13.753052   
15850  790475335086864240  491526222  Phakhamon   Ratchathewi   13.753169   
15851  790475546213717328  491526222  Phakhamon   Ratchathewi   13.754789   
15852  790476492384199044  491526222  Phakhamon   Ratchathewi   13.752960   

        longitude        room_type  price  minimum_nights  number_of_reviews  \
1      100.616740  Private room     1316   1               0                   
2      100.624020  Private room     800    60              0                   
4      100.633380  Private room     1905   1               0                   
7      100.567370  Entire home/apt  1461   1               0                   
14     100.568060  Entire home/apt  1400   30              0                   
...           ...              ...   ...   ..             ..                   
15848  100.653458  Private room     2298   28              0                   
15849  100.538738  Private room     1429   1               0                   
15850  100.538700  Private room     1214   1               0                   
15851  100.538757  Private room     1214   1               0                   
15852  100.540820  Private room     1214   1               0                   

      last_review  reviews_per_month  calculated_host_listings_count  \
1      -          NaN                 2                                
2      -          NaN                 1                                
4      -          NaN                 3                                
7      -          NaN                 1                                
14     -          NaN                 1                                
...   ..           ..                ..                                
15848  -          NaN                 1                                
15849  -          NaN                 14                               
15850  -          NaN                 14                               
15851  -          NaN                 14                               
15852  -          NaN                 14                               

       availability_365  number_of_reviews_ltm  
1      358               0                      
2      365               0                      
4      365               0                      
7      365               0                      
14     365               0                      
...    ...              ..                      
15848  362               0                      
15849  365               0                      
15850  365               0                  

In [144]:
#Proses mengubah missing value pada kolom `reviews_per_month` menjadi "0"
df['reviews_per_month'].fillna('0', inplace=True)
df[df['reviews_per_month'] == '0']

name  index  \
1      Easy going landlord,easy place                      1       
2      modern-style apartment in Bangkok                   2       
4      Suite Room 3 at MetroPoint                          4       
7      1 chic bedroom apartment in BKK                     7       
14     Deluxe Condo, Nana, Pool/GYM/Sauna                  14      
...                                   ...                  ..      
15848  素坤逸核心两房公寓42楼，靠近BTSon nut/无边天际泳池观赏曼谷夜景/出门当地美食街       15849   
15849  Euro LuxuryHotel PratunamMKt TripleBdNrShopingArea  15850   
15850  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15851   
15851  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15852   
15852  Euro LuxuryHotel PratunamMKt TwinBedNrShopingArea   15853   

                       id    host_id   host_name neighbourhood   latitude  \
1      27979               120541     Emy         Bang Na       13.668180   
2      28745               123784     Familyroom  Bang Kapi     13.752320   
4      941865              610315     Kasem       Bang Kapi     13.768720   
7      1738669             7045870    Jiraporn    Chatu Chak    13.829250   
14     959254              5153476    Natcha      Khlong Toei   13.715160   
...       ...                  ...       ...              ...         ...   
15848  790465040741092826  94899359   Renee       Pra Wet       13.715132   
15849  790474503157243541  491526222  Phakhamon   Ratchathewi   13.753052   
15850  790475335086864240  491526222  Phakhamon   Ratchathewi   13.753169   
15851  790475546213717328  491526222  Phakhamon   Ratchathewi   13.754789   
15852  790476492384199044  491526222  Phakhamon   Ratchathewi   13.752960   

        longitude        room_type  price  minimum_nights  number_of_reviews  \
1      100.616740  Private room     1316   1               0                   
2      100.624020  Private room     800    60              0                   
4      100.633380  Private room     1905   1               0                   
7      100.567370  Entire home/apt  1461   1               0                   
14     100.568060  Entire home/apt  1400   30              0                   
...           ...              ...   ...   ..             ..                   
15848  100.653458  Private room     2298   28              0                   
15849  100.538738  Private room     1429   1               0                   
15850  100.538700  Private room     1214   1               0                   
15851  100.538757  Private room     1214   1               0                   
15852  100.540820  Private room     1214   1               0                   

      last_review reviews_per_month  calculated_host_listings_count  \
1      -           0                 2                                
2      -           0                 1                                
4      -           0                 3                                
7      -           0                 1                                
14     -           0                 1                                
...   ..          ..                ..                                
15848  -           0                 1                                
15849  -           0                 14                               
15850  -           0                 14                               
15851  -           0                 14                               
15852  -           0                 14                               

       availability_365  number_of_reviews_ltm  
1      358               0                      
2      365               0                      
4      365               0                      
7      365               0                      
14     365               0                      
...    ...              ..                      
15848  362               0                      
15849  365               0                      
15850  365               0                      
15851  

> *Missing value* pada kolom `reviews_per_month` kini telah diisi oleh *value* "0"

##### 5. Identifikasi Unique Data

In [145]:
pd.set_option('display.max_colwidth', 1)

#Menampilkan data unik dari setiap kolom
listItem=[]
for col in df.columns:
    listItem.append([col, df[col].nunique(), df[col].unique()])

UniqueTable=pd.DataFrame(columns=['Nama Kolom', 'Jumlah Data Unik', 'Sample Unik'],
                        data=listItem)
UniqueTable

Nama Kolom  Jumlah Data Unik  \
0   name                            14794              
1   index                           15853              
2   id                              15853              
3   host_id                         6658               
4   host_name                       5312               
5   neighbourhood                   50                 
6   latitude                        9606               
7   longitude                       10224              
8   room_type                       4                  
9   price                           3039               
10  minimum_nights                  86                 
11  number_of_reviews               298                
12  last_review                     1670               
13  reviews_per_month               514                
14  calculated_host_listings_count  50                 
15  availability_365                366                
16  number_of_reviews_ltm           85                 

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

> - Setelah dilakukan pengetesan dan pengecekan ulang, kolom `ID` yang bertindak sebagai *primary key* sudah dipastikan tidak memiliki duplikasi data.
> - Kolom `room_type` memiliki 4 opsi *value*, yaitu Entire home/apt, Private room, Hotel room, dan Shared room.
> - Sudah tidak terdapat *missing value.*

Secara keseluruhan, pada tahapan *data understanding* ini bisa disimpulkan menjadi beberapa poin berikut:
1. Jumlah keseluruhan pada dataset ini adalah sebanyak 15854 data yang terbagi pada 17 kolom.
1. Format pada kolom `last_review` masih keliru, seharusnya berformat *date.*
1. Anomali data pada kolom `price` telah berhasil ditangani dengan melakukan *drop row data*.
1. Kolom `ID` yang bersifat sebagai *primary key* tidak memiliki duplikasi data.
1. *Missing value* pada kolom `name`, `host_name`, `last_review`, `reviews_per_month` sudah berhasil ditangani.
1. Kolom `room_type` memiliki 4 opsi *value*, yaitu Entire home/apt, Private room, Hotel room, dan Shared room.